Allow sharing the host PID namespace

pweil- opened the pull request:
In the same vein as https://github.com/kubernetes/kubernetes/pull/12470 this allows super privileged containers to share the host's PID namespace by adding
1.  A knob on the PodSpec to request enabling the feature
2.  A sources list flag in the kubelet for allowable sources with checks in `canRunPod`
3.  Docker support for setting the config and a TODO in rkt
4.  An admission controller to prevent exec into a container with HostPID enabled

@bgrant0607 - requested API change
@liggitt @smarterclayton @erictune @simon3z @pmorie - PTAL
@derekwaynecarr - admission addition

pmorie commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -1497,7 +1499,7 @@ func (dm *DockerManager) createPodInfraContainer(pod *api.Pod) (kubeletTypes.Doc
 	}
 
-	id, err := dm.runContainerInPod(pod, container, netNamespace, "")
+	id, err := dm.runContainerInPod(pod, container, netNamespace, "", "")
```
Shouldn't pidMode be passed to runContainerInPod here?

pweil- replied:
It could but I don't think it's strictly required.  Since the ipc mode wasn't being passed I did the same.

pmorie replied:
There were some further comments about that.  I think the pid mode has to be passed in, IIUC

pmorie replied:
https://github.com/kubernetes/kubernetes/pull/12470#issuecomment-136860026

pweil- replied:
Added pid mode to infra container and moved the logic produce the pid mode string to a method and corresponding test case since it is used in two spots now.

bgrant0607 commented:
cc @kubernetes/goog-node

ltagliamonte commented:
Hello any ETA on when this will be merged?

pmorie commented on pkg/api/v1/types.go:
```diff
@@ -1229,6 +1229,9 @@ type PodSpec struct {
 	HostNetwork bool `json:"hostNetwork,omitempty"`
+	// Uses the host's pid namespace.
+	// Optional: Default to false.
+	HostPID bool `json:"hostPID,omitempty"`
```
Will you make this a pointer to distinguish unset values?

pmorie replied:
Nvm, I don't think I need this anymore.

derekwaynecarr commented:
Admission control piece looks fine to me. 

@bgrant0607 - please ack the API change.

bgrant0607 commented:
I'm fine with the API change. It would be good to get closure on #12470 as well.

cc @vishh on the PID namespace change.

bgrant0607 commented on pkg/api/v1/types.go:
```diff
@@ -1229,6 +1229,9 @@ type PodSpec struct {
 	// If this option is set, the ports that will be used must be specified.
 	// Default to false.
 	HostNetwork bool `json:"hostNetwork,omitempty"`
+	// Uses the host's pid namespace.
```
Use the host's PID namespace.  (rather than "Uses")

Please make the comment for HostNetwork consistent with this, also: Use the host's network namespace. (rather than "Uses")

liggitt commented on plugin/pkg/admission/exec/denyhostpid/admission.go:
```diff
@@ -0,0 +1,67 @@
+)
+
+func init() {
+	admission.RegisterPlugin("DenyExecOnHostPID", func(client client.Interface, config io.Reader) (admission.Interface, error) {
```
we're going to need a zillion of these... at the very least, we'd want a single controller with flags like this:
```
LimitExec
LimitAttach

LimitHostIPC
LimitHostPID
LimitPrivileged
```
so we can quickly register granular admission controllers, as well as a single one that rolls up all of them (which is what we'd probably want to use in practice)

liggitt replied:
See admission changes in https://github.com/kubernetes/kubernetes/pull/13705/files#diff-1d1d25a869a428f378f5ddd0d7943706

vishh commented:
kubelet changes LGTM

derekwaynecarr commented:
rebase please

pweil- commented:
rebased.  Updated type comments from "uses" to "use" for both pid mode and host network.  Updated admission to include the `pods/attach` subresource.  Agree that we should create a compound admission controller for these but probably as a separate pull. 

PTAL, will squash when I get the thumbs up

liggitt commented:
can we wait to add the new admission controller name to all the default lists and just do the compound as a quick follow up? remember that every admission controller name we add is an exposed name that will make it into config scripts, which means we should make every effort to keep support that name

pweil- commented:
> can we wait to add the new admission controller name to all the default lists and just do the compound as a quick follow up?

Sure, I don't mind.  Seems like the individual names will still need to make it into at least the  docs/help pieces though if users have the option of using them individually rather than the roll up version.

@derekwaynecarr ?

derekwaynecarr commented:
@pweil- Is there ever a deployment scenario where we would want to allow Exec on HostPID?  Is this something that should always just be denied and not require a user configuration piece to enable the denial?  For example, should we start to have some default admission controllers that always run at this point implicitly without user configuration?  I am starting to empathize with the deployer with the proliferation of flag settings needed, and share similar concerns to @liggitt above.  How many different controllers do we anticipate growing to that we would then consolidate into a single compound best practice admission controller?

derekwaynecarr commented:
@pweil- or to phrase another way, I am fine with a **single** admission controller that we grow in the future that sets all the proper flags to protect the host, and that we use that as a model, instead of requiring multiple individual controllers to be added to provide a locked down host each tweaking their own knobs.   If you want a different set of knobs to allow for some other variation of security, then write a different admission controller that makes that clear.  The piecemeal approach is starting to feel like an easy way to misconfigure the system and introduce problems.  let's just have a single plug-in for now.

pmorie commented:
@pweil- #12823 is going to move this field -- should we wait until that goes in, or just move the field when it goes in, since we won't have released a client that uses this field yet?

pweil- commented:
> @pweil- #12823 is going to move this field -- should we wait until that goes in, or just move the field when it goes in, since we won't have released a client that uses this field yet?

I'm hoping to finish the new roll-up admission controller on Monday.  If #12823 is going to go in soon then it can wait, if not I'd like this to go in.

pmorie commented:
I want #12823 to go in monday if possible.

On Sunday, September 13, 2015, Paul Weil <notifications@github.com> wrote:

> @pweil- <https://github.com/pweil-> #12823
> <https://github.com/kubernetes/kubernetes/pull/12823> is going to move
> this field -- should we wait until that goes in, or just move the field
> when it goes in, since we won't have released a client that uses this field
> yet?
>
> I'm hoping to finish the new roll-up admission controller on Monday. If
> #12823 <https://github.com/kubernetes/kubernetes/pull/12823> is going to
> go in soon then it can wait, if not I'd like this to go in.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13447#issuecomment-139921995>
> .
>

pweil- commented:
@derekwaynecarr @liggitt @vishh - updated to use a single admission controller that supports pid, ipc, and privileged denials.  PTAL

smarterclayton commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -1889,3 +1891,12 @@ func (dm *DockerManager) doBackOff(pod *api.Pod, container *api.Container, podSt
+func getPidMode(pod *api.Pod) string {
+	pidMode := ""
+	if pod.Spec.HostPID {
+		pidMode = "host"
```
SHouldn't this be a constant?  Or is this docker specific and the client doesn't offer a constant?

pweil- replied:
I did not see a constant for this in the docker client packages.  UTS mode and host network also are using `"host"` in `manager.go` so I went with that approach.

pmorie replied:
@smarterclayton if we can get #12823 in I woukd really like this to go in
afterward.

On Monday, September 14, 2015, Paul Weil <notifications@github.com> wrote:

> In pkg/kubelet/dockertools/manager.go
> <https://github.com/kubernetes/kubernetes/pull/13447#discussion_r39417286>
> :
>
> > @@ -1889,3 +1891,12 @@ func (dm *DockerManager) doBackOff(pod *api.Pod, container *api.Container, podSt
> >  	dm.clearReasonCache(pod, container)
> >  	return false
> >  }
> > +
> > +// getPidMode returns the pid mode to use on the docker container based on pod.Spec.HostPID.
> > +func getPidMode(pod *api.Pod) string {
> > +	pidMode := ""
> > +	if pod.Spec.HostPID {
> > +		pidMode = "host"
>
> I did not see a constant for this in the docker client packages. UTS mode
> and host network also are using "host" in manager.go so I went with that
> approach.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13447/files#r39417286>.
>

smarterclayton commented:
One comment then this LGTM - we have API sign off and Vish's Kubelet sign off and Derek's, so I'll put the tag on after that.

pweil- commented:
need to reconcile this comment https://github.com/kubernetes/kubernetes/pull/12470#issuecomment-140140896 before tagging.  May need to wait and rebase on @pmorie's PR if having this on the pod spec is undesirable.

bgrant0607 commented:
I think we should move forward with this PR.

derekwaynecarr commented on plugin/pkg/admission/exec/admission.go:
```diff
@@ -0,0 +1,101 @@
+		return admission.NewForbidden(a, fmt.Errorf("Cannot exec into or attach to a container using host pid"))
+	}
+
+	//TODO uncomment when this lands
```
is there an issue that we can reference to track this?

pweil- replied:
Added issue link to TODO.  Also already commented about the new admission strategy and this code on https://github.com/kubernetes/kubernetes/pull/12470#issuecomment-140106749

derekwaynecarr commented:
LGTM pending question on // TODO

liggitt commented on plugin/pkg/admission/exec/admission.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+func init() {
+	admission.RegisterPlugin("DenyExec", func(client client.Interface, config io.Reader) (admission.Interface, error) {
```
Probably needs a slightly tweaked name. It doesn't deny exec, it denies "dangerous" or "escalating" exec

liggitt replied:
I'd also like to see the existing deny privileged exec controller go away and register a construction of this one under the old name with the correct flags to get the old behavior

liggitt replied:
And we should probably replace DenyPrivilegedExec with the name of this combined one (once that's decided). @derekwaynecarr do you agree this rolled up one is what most people should be using

pweil- commented:
@liggitt @derekwaynecarr - renamed to `DenyEscalatingExec` and removed the old controller.  Registering the old controller's name with tweaked settings in the `DenyEscalatingExec` init method now.  (2nd commit)

derekwaynecarr commented on docs/admin/kube-apiserver.md:
````diff
@@ -48,7 +48,7 @@ cluster's shared state through which all other components interact.
 ```
       --address=<nil>: DEPRECATED: see --insecure-bind-address instead
-      --admission-control="": Ordered list of plug-ins to do admission control of resources into cluster. Comma-delimited list of: AlwaysAdmit, AlwaysDeny, DenyExecOnPrivileged, LimitRanger, NamespaceAutoProvision, NamespaceExists, NamespaceLifecycle, ResourceQuota, SecurityContextDeny, ServiceAccount
+      --admission-control="": Ordered list of plug-ins to do admission control of resources into cluster. Comma-delimited list of: AlwaysAdmit, AlwaysDeny, DenyEscalatingExec, LimitRanger, NamespaceAutoProvision, NamespaceExists, NamespaceLifecycle, ResourceQuota, SecurityContextDeny, ServiceAccount
````
This md document should still referenced the deprecated option for ```DenyExecOnPrivileged``` since its still valid input.

derekwaynecarr commented:
Few more items:

* Mark ```DenyExecOnPrivileged``` deprecated here https://github.com/kubernetes/kubernetes/blob/master/docs/admin/admission-controllers.md#denyexeconprivileged
* Document your new controller in that document explaining what it prevents.

pweil- commented:
@derekwaynecarr docs updated.  If it looks good I'll squash

smarterclayton commented:
Docs look fine, squash

Commit:
pid mode

vishh commented on docs/admin/kubelet.md:
```diff
@@ -84,6 +84,7 @@ HTTP server: The kubelet can also listen for HTTP and respond to a simple API
       --healthz-port=0: The port of the localhost healthz endpoint
   -h, --help=false: help for kubelet
       --host-network-sources="": Comma-separated list of sources from which the Kubelet allows pods to use of host network. For all sources use "*" [default="file"]
+      --host-pid-sources="": Comma-separated list of sources from which the Kubelet allows pods to use the host pid namespace. For all sources use "*" [default="file"]
```
Do we need one flag per namespace or can we combine them? Do we have use cases where admins will enable host namespace sharing for individual namespaces?
cc @smarterclayton

smarterclayton commented:
When PodSecurityPolicy lands that covers the individual use cases.  I'd
prefer not to make the default one that flexible (because PSP is intended
to solve this problems holistically).

On Tue, Sep 15, 2015 at 1:46 PM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *failed* for commit c61a8bd
> <https://github.com/kubernetes/kubernetes/commit/c61a8bdc4a45c42385b200cea6e4dd7909edaa29>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/c61a8bdc4a45c42385b200cea6e4dd7909edaa29/kubernetes-pull-build-test-e2e-gce/8559/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/c61a8bdc4a45c42385b200cea6e4dd7909edaa29/kubernetes-pull-build-test-e2e-gce/8559/_artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//8559>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13447#issuecomment-140479553>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

pweil- commented:
> When PodSecurityPolicy lands that covers the individual use cases.  I'd
> prefer not to make the default one that flexible (because PSP is intended
> to solve this problems holistically).

@vishh - I believe that was in response to your question.  

Any other feedback on this?

vishh commented:
Acknowledged. No other comments.

On Tue, Sep 15, 2015 at 12:01 PM, Paul Weil <notifications@github.com>
wrote:

> When PodSecurityPolicy lands that covers the individual use cases. I'd
> prefer not to make the default one that flexible (because PSP is intended
> to solve this problems holistically).
>
> @vishh <https://github.com/vishh> - I believe that was in response to
> your question.
>
> Any other feedback on this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13447#issuecomment-140502546>
> .
>

derekwaynecarr commented:
LGTM - let's do this

derekwaynecarr commented:
@k8s-bot add to whitelist

derekwaynecarr commented:
@k8s-bot test this

liggitt commented:
I think the merge whitelist comes from https://github.com/kubernetes/contrib/blob/master/submit-queue/whitelist.txt, and the bot whitelist command is for testing

alex-mohr commented:
@k8s-bot test this again please
(previous failures: Kubernetes e2e suite.Daemon set should launch a daemon pod on every node of the cluster
junit_03.xml.<init>)
